src/ModelContextProtocol.AspNetCore/McpRouteAwarenessExtension.cs

```diff
@@ -0,0 +1,228 @@
+
+                // If no route service or requesting the global route, serve all tools
+                if (routeService?.GlobalRoute is null || string.Equals(requestPath, routeService.GlobalRoute, StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                var toolCollection = mcpOptions.Capabilities?.Tools?.ToolCollection;
+                if (toolCollection is null)
+                {
+                    return;
+                }
+
+                // Create a snapshot of current tools before filtering
+                var allTools = toolCollection.ToList();
+                toolCollection.Clear();
+
+                // Filter tools based on the requested route
+                foreach (var tool in allTools)
+                {
+                    if (ShouldIncludeTool(tool, requestPath, routeService))
+                    {
+                        toolCollection.Add(tool);
+                    }
+                }
+
+                // Note: Resources and prompts are not filtered - they remain available on all routes
+            };
+
+            // Apply any additional configuration
+            configureOptions?.Invoke(options);
+        });
+
+        // Register the route service as a singleton
+        builder.Services.AddSingleton<RouteAwareToolService>();
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Maps MCP endpoints with route-aware capabilities, creating endpoints for both the global route and all discovered tool routes.
+    /// </summary>
+    /// <param name="endpoints">The endpoint route builder to configure.</param>
+    /// <param name="pattern">The route pattern prefix for the global MCP endpoint.</param>
+    /// <returns>An <see cref="IEndpointConventionBuilder"/> for configuring additional endpoint conventions.</returns>
+    /// <remarks>
+    /// <para>
+    /// This method discovers all tool routes from registered <see cref="McpServerTool"/> instances
+    /// and creates MCP endpoints for each unique route. Tools without route attributes are
+    /// available on the global route and all specific routes.
+    /// </para>
+    /// <para>
+    /// For example, with a global pattern of "mcp" and tools having routes "admin" and "readonly":
+    /// <list type="bullet">
+    /// <item><description>/mcp - serves all tools (global route)</description></item>
+    /// <item><description>/mcp/admin - serves admin-specific tools + global tools</description></item>
+    /// <item><description>/mcp/readonly - serves readonly-specific tools + global tools</description></item>
+    /// </list>
+    /// </para>
+    /// </remarks>
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="endpoints"/> is null.</exception>
+    public static IEndpointConventionBuilder MapMcpWithRouting(this IEndpointRouteBuilder endpoints, [StringSyntax("Route")] string pattern = "")
```

Comment:
I don't think we should add a new `MapMcpWithRouting` method. If we want to support routing via attributes, we should support that via the normal `MapMcp` method.